rfcs/0154-Migrate-queue-to-postgres.md

```diff
@@ -0,0 +1,102 @@
+# RFC 154 - Migrate Taskcluster to postgres
+* Updates: RFC [65](https://github.com/taskcluster/taskcluster-rfcs/blob/master/rfcs/0065-Migrate-queue-to-postgres.md)
+* Obsoletes: None
+* Comments: [#154](https://github.com/taskcluster/taskcluster-rfcs/pull/154)
+* Initially Proposed by: @jonasfj
+
+# Summary
+
+Taskcluster uses Azure as the data-storage backend. For example, the Queue
+service currently uses a bunch of Azure tables to handle tasks, dependencies,
+artifacts, and so on. This project involves refactoring Taskcluster to use a
+Postgres backend, instead.
+
+# Motivation
+
+* A shared base of indexed data can allow services to make smarter decisions and
+  save money on provisioning and data storage
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures
+* Storing data in postgres will enable features users desire:
+  * Queue introspection
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status
+    * What tasks has this worker performed?
+  * Ad-hoc analytics
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically changing task levels
+
+# Details
+
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+## Schema Migration
+
+Taskcluster will write its own tooling to handle schema migrations and will have
+all migrations tested with sample data to ensure we get the expected schema. In
+addition, there will be one schema for the whole monorepo, so that schema
+updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades Compatibility
```

Comment:
> The approach I'm familiar with for the versioning problem is to follow the sequence of changes below

Doing it that way would require the deployer to perform 3 tasks when upgrading which puts the burden on the deployer. We are trying to make Taskcluster a product used outside Mozilla and so whenever possible, we'd like to remove the burden from the deployer. Using stored procedures would allow us to make changes to the internal representation of the database without causing breaking changes to consumers of the db.